File: 프로그래머스/1/340198. ［PCCE 기출문제］ 10번 ／ 공원/［PCCE 기출문제］ 10번 ／ 공원.py
def check(park, mat, i, j):
    for x in range(i, i + mat):
        for y in range(j, j + mat):
            if x >= n or y >= m or park[x][y] != '-1':
                return 0
    return 1

def spread(park, mat):
    for i in range(n):
        for j in range(m):
            if park[i][j] != '-1':
                continue
            if check(park, mat, i, j):
                answer.append(mat)
                return

def solution(mats, park):
    global n, m, answer
    
    n, m = len(park), len(park[0])
    
    answer = []
    for mat in mats:
        spread(park, mat)
    return max(answer) if answer else -1

File: 프로그래머스/1/340198. ［PCCE 기출문제］ 10번 ／ 공원/［PCCE 기출문제］ 10번 ／ 공원.java
class Solution {
    
    int n, m, answer;
    
    public int solution(int[] mats, String[][] park) {
        n = park.length; m = park[0].length;

        answer = -1;
        for (int mat : mats) {
            spread(park, mat);
        }
        return answer;
    }
    
    public boolean spread(String[][] park, int mat) {
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (!park[i][j].equals("-1")) continue;
                if (check(park, mat, i, j)) {
                    answer = Math.max(answer, mat);
                    return true;
                }
            }
        }
        return false;
    }
    
    public boolean check(String[][] park, int mat, int i, int j) {
        for (int x = i; x < i + mat; x++) {
            for (int y = j; y < j + mat; y++) {
                if (x >= n || y >= m || !park[x][y].equals("-1")) {
                    return false;
                }    
            }
        }
        return true;
    }
    
}